Public/Src/Cache/ContentStore/Library/Stores/FileSystemContentStoreInternal.cs

```diff
@@ -349,6 +354,195 @@ private Task Upgrade2To3(Context context)
+                {
+                    return (isValid: false, error: $"File size mismatch. Expected size is {contentFileInfo.FileSize} and size on disk is {fileInfo.Length}.");
+                }
+
+                // Or if the content doesn't match the hash.
+                var actualHashAndSize = await TryHashFileAsync(context, fileInfo.FullPath, expectedHash.HashType);
+                if (actualHashAndSize != null && actualHashAndSize.Value.Hash != expectedHash)
+                {
+                    // Don't need to add an expected hash into the error string because the client code will always put it into the final error message.
+                    return (isValid: false, error: $"Hash mismatch. Actual hash is {actualHashAndSize.Value.Hash}.");
+                }
+
+                return (isValid: true, error: string.Empty);
+            }
+            catch (IOException e)
+            {
+                _tracer.Warning(context, $"Content hash validation failed. Hash={expectedHash}, Error={e}.");
+
+                return (isValid: true, error: string.Empty);
+            }
+        }
+
+        private async Task<ContentHashWithSize?> TryHashFileAsync(Context context, AbsolutePath path, HashType hashType, Func<Stream, Stream> wrapStream = null)
+        {
+            // We only hash the file if a trusted hash is not supplied
+            using (var stream = await FileSystem.OpenAsync(path, FileAccess.Read, FileMode.Open, FileShare.Read | FileShare.Delete))
+            {
+                if (stream == null)
+                {
+                    return null;
+                }
+
+                using (var wrappedStream = (wrapStream == null) ? stream : wrapStream(stream))
+                {
+                    // Hash the file in  place
+                    return await HashContentAsync(context, wrappedStream, hashType, path);
+                }
+            }
+        }
+
+        private bool SelfCheckUpToDate()
+        {
+            if (!FileSystem.FileExists(_selfCheckFilePath))
+            {
+                return false;
+            }
+
+            // The format is: "SelfCheckEpoch|LastSelfCheckTime"
+            var contents = FileSystem.ReadAllText(_selfCheckFilePath);
+            var parts = contents.Split('|');
+            if (parts.Length != 2 || parts[0] != _settings.SelfCheckEpoch)
+            {
+                return false;
+            }
+
+            var reconcileTime = DateTimeUtilities.FromReadableTimestamp(parts[1]);
+            if (reconcileTime == null)
+            {
+                return false;
+            }
+
+            return reconcileTime.Value.IsRecent(Clock.UtcNow, _settings.SelfCheckFrequency);
+        }
+
+        private void MarkSelfChecked(bool success)
+        {
+            if (success)
+            {
+                FileSystem.WriteAllText(_selfCheckFilePath, $"{_settings.SelfCheckEpoch}|{Clock.UtcNow.ToReadableString()}");

```

Comment:
>WriteAllText [](start = 27, length = 12)

Include list of paths which were scanned to skip when resuming check.

Reply:
As discussed, will only trace the last hash.

---
In reply to: [282936721](https://github.com/microsoft/BuildXL/pull/289#discussion_r282936721) [](ancestors = 282936721)